R/select_utilities.R

```diff
@@ -0,0 +1,275 @@
+    stop("Invalid syntax", call. = FALSE)
+  }
+
+  # selecting the variable from the variable information data frame
+  env_variable_type$df_var_info %>%
+    dplyr::select(all_of(c(variable_column, select_column))) %>%
+    dplyr::filter(stats::complete.cases(.)) %>%
+    dplyr::filter({{ select_expr }}) %>%
+    dplyr::pull(all_of(variable_column)) %>%
+    unique()
+}
+
+
+# scoping the variable characteristics
+.scope_var_info <- function(x) {
+  # removing everything from selecting environment
+  rm(list = ls(envir = env_variable_type), envir = env_variable_type)
+  if (!inherits(x, "data.frame")) return(invisible(NULL))
+
+  # saving list of variable types to selecting environment
+  df_var_info <-
+    x %>%
+    dplyr::select(any_of(c("variable", "var_label", "var_class",
+                           "var_type", "var_nlevels", "contrasts"))) %>%
+    dplyr::distinct()
+
+  env_variable_type$df_var_info <- df_var_info
+
+  return(invisible(NULL))
+}
+
+# function that converts a meta_data tibble to a tibble of variable names (to be used in selecting)
+.var_info_to_df <- function(x) {
+  # converting variable name and class into data frame so users can use `where(predicate)`-types
+  if (inherits(x, "data.frame") && all(c("variable", "var_class") %in% names(x))) {
+    # keep unique var names
+    x <-
+      dplyr::select(x, all_of(c("variable", "var_class"))) %>%
+      dplyr::distinct() %>%
+      dplyr::filter(!is.na(.data$variable))
+    df <-
+      purrr::map2_dfc(
+        x$variable, x$var_class,
+        function(var, class) {
+          switch(
+            class,
+            "numeric" = data.frame(pi),
+            "character" = data.frame(letters[1]),
+            "factor" = data.frame(datasets::iris$Species[1]),
+            "ordered" = data.frame(factor(datasets::iris$Species[1], ordered = TRUE)),
+            "integer" = data.frame(1L),
+            "Date" = data.frame(Sys.Date()),
+            "difftime" = data.frame(Sys.Date() - Sys.Date())
+          ) %||%
+            data.frame(NA) %>%
+            purrr::set_names(var)
+        }
+      )
+  }
+  # if a data.frame
+  else if (inherits(x, "data.frame") && "variable" %in% names(x)) {
+    df <- purrr::map_dfc(unique(x$variable), ~data.frame(NA) %>% purrr::set_names(.x))
+  }
+  # if only a vector of names were passed, converting them to a data frame
+  else if (rlang::is_vector(x) && !is.list(x)) {
+    df <- purrr::map_dfc(unique(x), ~data.frame(NA) %>% purrr::set_names(.x))
+  }
+  # return data frame with variables as column names
+  df
+}
+
+# extract LHS/RHS of formula as quosure. attached env will be the formula env
+.f_side_as_quo <- function(x, side = c("lhs", "rhs")) {
+  side <- match.arg(side)
+  f_expr <-
+    switch(side,
+           "lhs" = rlang::f_lhs(x),
+           "rhs" = rlang::f_rhs(x))
+  f_quo <- rlang::quo(!!f_expr)
+  attr(f_quo, ".Environment") <- rlang::f_env(x)
+  f_quo
+}
+
+formula_select_error <- function(arg_name) {
```

Comment:
as an internal function should it be prefixed by a dot?

Reply:
and add two lines of comments explaining the function

Reply:
✔️